pandas/core/arrays/boolean.py

```diff
@@ -0,0 +1,679 @@
+
+from pandas.core import nanops, ops
+from pandas.core.algorithms import take
+from pandas.core.arrays import ExtensionArray, ExtensionOpsMixin
+
+if TYPE_CHECKING:
+    from pandas._typing import Scalar
+
+
+@register_extension_dtype
+class BooleanDtype(ExtensionDtype):
+    """
+    Extension dtype for boolean data.
+
+    .. versionadded:: 1.0.0
+
+    .. warning::
+
+       BooleanDtype is considered experimental. The implementation and
+       parts of the API may change without warning.
+
+    Attributes
+    ----------
+    None
+
+    Methods
+    -------
+    None
+
+    Examples
+    --------
+    >>> pd.BooleanDtype()
+    BooleanDtype
+    """
+
+    @property
+    def na_value(self) -> "Scalar":
+        """
+        StringDtype uses :attr:`numpy.nan` as the missing NA value.
+
+        .. warning::
+
+           `na_value` may change in a future release.
+        """
+        return np.nan
+
+    @property
+    def type(self) -> Type:
+        return np.bool_
+
+    @property
+    def kind(self) -> str:
+        return "b"
+
+    @property
+    def name(self) -> str:
+        """
+        The alias for StringDtype is ``'string'``.
+        """
+        return "boolean"
+
+    @classmethod
+    def construct_from_string(cls, string: str) -> ExtensionDtype:
+        if string == "boolean":
+            return cls()
+        return super().construct_from_string(string)
+
+    @classmethod
+    def construct_array_type(cls) -> "Type[BooleanArray]":
+        return BooleanArray
+
+    def __repr__(self) -> str:
+        return "BooleanDtype"
+
+    @property
+    def _is_boolean(self) -> bool:
+        return True
+
+
+def coerce_to_array(values, mask=None, copy=False):
+    """
+    Coerce the input values array to numpy arrays with a mask
+
+    Parameters
+    ----------
+    values : 1D list-like
+    dtype : integer dtype
+    mask : bool 1D array, optional
+    copy : bool, default False
+        if True, copy the input
+
+    Returns
+    -------
+    tuple of (values, mask)
+    """
+    if isinstance(values, BooleanArray):
+        if mask is not None:
+            raise ValueError("cannot pass mask for BooleanArray input")
+        values, mask = values._data, values._mask
+        if copy:
+            values = values.copy()
+            mask = mask.copy()
+        return values, mask
+
+    mask_values = None
+    if isinstance(values, np.ndarray) and values.dtype == np.bool_:
+        if copy:
+            values = values.copy()
+    else:
+        values_object = np.asarray(values, dtype=object)
+        mask_values = isna(values)
+        values = np.zeros(len(values), dtype=bool)
+        values[~mask_values] = values_object[~mask_values].astype(bool)
+
+    if isinstance(mask, np.ndarray) and mask.dtype == np.bool_:
+        if copy:
+            mask = mask.copy()
+    else:
+        if mask is None:
+            if mask_values is None:
+                # TODO optimize this case?
```

Comment:
Can you clarify the potential optimization in the comment?

Reply:
that's a left-over from copying it from integer, can probably be removed
It might be related to the general optimization that we can do for those mask-based arrays to allow a None mask if there are no missing values.